src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,224 @@
+
+#include "ray/core_worker/transport/direct_actor_transport.h"
+#include "ray/common/task/task.h"
+
+using ray::rpc::ActorTableData;
+
+namespace ray {
+
+bool HasByReferenceArgs(const TaskSpecification &spec) {
+  for (int i = 0; i < spec.NumArgs(); ++i) {
+    if (spec.ArgIdCount(i) > 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+CoreWorkerDirectActorTaskSubmitter::CoreWorkerDirectActorTaskSubmitter(
+    boost::asio::io_service &io_service,
+    gcs::GcsClient &gcs_client,
+    CoreWorkerObjectInterface &object_interface)
+    : io_service_(io_service),
+      gcs_client_(gcs_client),
+      client_call_manager_(io_service),
+      store_provider_(object_interface.CreateStoreProvider(StoreProviderType::LOCAL_PLASMA)),
+      counter_(0) {
+       
+  RAY_CHECK_OK(SubscribeActorTable());
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(const TaskSpec &task) {
+
+  if (HasByReferenceArgs(task.GetTaskSpecification())) {
+    return Status::Invalid("direct actor call only supports by value arguments");
+  }
+
+  RAY_CHECK(task.GetTaskSpecification().IsActorTask());
+  const auto &actor_id = task.GetTaskSpecification().ActorId();
+
+  const auto task_id = task.GetTaskSpecification().TaskId();
+  const auto num_returns = task.GetTaskSpecification().NumReturns();
+
+  auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
+  request->set_task_id(task.GetTaskSpecification().TaskId().Binary());
+  request->set_task_spec(task.GetTaskSpecification().Serialize());
+
+  std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+  auto entry = rpc_clients_.find(actor_id);
+  if (entry == rpc_clients_.end()) {
+    // TODO: what if actor is not created yet?
+    // if (pending_requests_[actor_id].empty()) {
+    //   gcs_client_.Actors().RequestNotifications(JobID::Nil(), actor_id,
+    //      gcs_client_.GetLocalClientID());
+    // }
+
+    auto pending_request =
+        std::unique_ptr<PendingTaskRequest>(new PendingTaskRequest);
+    pending_request->task_id = task_id;
+    pending_request->num_returns = num_returns;
+    pending_request->request = std::move(request);
+    // append the task to the pending list.
+
+    pending_requests_[actor_id].emplace_back(std::move(pending_request));
+
+    return Status::OK();
+  }
+
+  auto iter = actor_state_.find(actor_id);
+  RAY_CHECK(iter != actor_state_.end());
+  if (iter->second != ActorTableData::ALIVE) {
```

Comment:
if actor state is reconstructing, should we cache this task and try later?

Reply:
I think that is also doable, and might be more desired